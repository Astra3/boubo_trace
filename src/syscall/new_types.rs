use libc::sockaddr;
use nix::{
    fcntl::OFlag,
    sys::{
        socket::{self, AddressFamily},
        stat::Mode,
    },
};
use serde::{Serialize, ser::SerializeStruct};

#[derive(Debug, Clone)]
#[allow(non_camel_case_types)]
pub struct sockaddr_ser(sockaddr);

impl Serialize for sockaddr_ser {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut sock_addr = serializer.serialize_struct("sockaddr", 2)?;
        sock_addr.serialize_field("sa_family", &self.0.sa_family)?;
        sock_addr.serialize_field("sa_data", &self.0.sa_data)?;
        sock_addr.end()
    }
}

impl From<sockaddr> for sockaddr_ser {
    fn from(val: sockaddr) -> Self {
        sockaddr_ser(val)
    }
}

#[derive(Debug, Clone)]
pub struct SocketType {
    r#type: socket::SockType,
    flags: socket::SockFlag,
}

impl Serialize for SocketType {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut sock_type = serializer.serialize_struct("socket_type", 2)?;
        sock_type.serialize_field("type", &(self.r#type as i32))?;
        sock_type.serialize_field("flags", &self.flags.bits())?;
        sock_type.end()
    }
}

impl From<i32> for SocketType {
    fn from(val: i32) -> Self {
        let flags = socket::SockFlag::from_bits_truncate(val);
        let sock_type = socket::SockType::try_from(val & !socket::SockFlag::all().bits()).unwrap();

        SocketType {
            r#type: sock_type,
            flags,
        }
    }
}

#[derive(Debug, Clone)]
pub struct AddressFamilySer(AddressFamily);

impl Serialize for AddressFamilySer {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_i32(self.0 as i32)
    }
}

impl From<AddressFamily> for AddressFamilySer {
    fn from(val: AddressFamily) -> Self {
        AddressFamilySer(val)
    }
}

#[derive(Debug, Clone)]
pub struct OFlagSer(OFlag);

impl Serialize for OFlagSer {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_i32(self.0.bits())
    }
}

impl From<OFlag> for OFlagSer {
    fn from(val: OFlag) -> Self {
        OFlagSer(val)
    }
}

#[derive(Debug, Clone)]
pub struct ModeSer(Mode);

impl Serialize for ModeSer {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_u32(self.0.bits())
    }
}

impl From<Mode> for ModeSer {
    fn from(val: Mode) -> Self {
        ModeSer(val)
    }
}
