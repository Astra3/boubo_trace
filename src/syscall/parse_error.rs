use nix::{
    errno::Errno,
    sys::{signal::Signal, wait::WaitStatus},
};
use serde::Serialize;
use thiserror::Error;

use super::SyscallDisc;

#[derive(Error, Debug)]
pub enum SyscallParseError {
    #[error("error '{error:?}' on syscall {syscall}")]
    SyscallError { syscall: SyscallDisc, error: Errno },
    #[error("error in syscall by tracer: {0:?}")]
    PtraceError(#[from] Errno),
    #[error("tracee process is not running and exited with status code {0}")]
    ProcessExit(i32),
    #[error("unexpected status returned by waitpid: {0:?}")]
    UnexpectedWaitStatus(WaitStatus),
    #[error("error returned by waitpid with errno: {0:?}")]
    WaitPidError(Errno),
    // FIXME this isn't really a parse error
    #[error("tracee terminated by OS with signal {signal:?}")]
    Terminated { signal: Signal, core_dumped: bool },
}

impl Serialize for SyscallParseError {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}
