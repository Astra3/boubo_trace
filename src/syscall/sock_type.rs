use nix::sys::socket;

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub struct SocketType {
    r#type: socket::SockType,
    flags: socket::SockFlag,
}

impl From<i32> for SocketType {
    fn from(value: i32) -> Self {
        let flags = socket::SockFlag::from_bits_truncate(value);
        let sock_type = socket::SockType::try_from(value & !socket::SockFlag::all().bits()).unwrap();

        SocketType { r#type: sock_type, flags }


    }
}
