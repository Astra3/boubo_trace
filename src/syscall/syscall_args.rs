use libc::user_regs_struct;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct SyscallArgs(
    pub(super) usize,
    pub(super) usize,
    pub(super) usize,
    pub(super) usize,
    pub(super) usize,
    pub(super) usize,
);

impl SyscallArgs {
    pub fn new(regs: &user_regs_struct) -> SyscallArgs {
        SyscallArgs(
            regs.rdi as usize,
            regs.rsi as usize,
            regs.rdx as usize,
            regs.r10 as usize,
            regs.r8 as usize,
            regs.r9 as usize,
        )
    }
}
