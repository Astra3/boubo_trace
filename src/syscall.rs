use libc::mode_t;
use log::{debug, error, trace, warn};
use nix::{
    errno::Errno,
    sys::{ptrace, wait::WaitStatus},
    unistd::Pid,
};
use serde::{Deserialize, Serialize};
use thiserror::Error;
use utils::{get_return_value, parse_return, read_bytes_memory, read_cstring, wait_for_stop};

mod utils;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, strum::EnumIs, strum::EnumDiscriminants)]
#[strum_discriminants(derive(strum::Display))]
#[serde(rename_all = "snake_case")]
pub enum Syscall {
    Read {
        fd: i32,
        read_bytes: Vec<u8>,
        requested_count: usize,
    },
    Write {
        fd: i32,
        buf: Vec<u8>,
    },
    Close {
        fd: i32,
    },
    Openat {
        dirfd: i32,
        pathname: Vec<u8>,
        flags: i32,
        mode: mode_t,
    },
    ExitGroup {
        status: i32,
    },
    Unknown {
        id: u64,
        args: SyscallArgs,
        return_value: i64,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct SyscallArgs(u64, u64, u64, u64, u64, u64);

type SyscallDisc = SyscallDiscriminants;

#[derive(Error, Debug)]
pub enum SyscallParseError {
    #[error("error in tracee's syscall")]
    SyscallError { syscall: SyscallDisc, error: Errno },
    #[error("error in syscall by tracer: {0:?}")]
    PtraceError(#[from] Errno),
    #[error("tracee process is not running")]
    ProcessExit(i32),
    #[error("unexpected status returned by waitpid")]
    UnexpectedWaitStatus(WaitStatus),
    #[error("error returned by waitpid")]
    WaitPidError(Errno),
}

impl Serialize for SyscallParseError {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer {
        serializer.serialize_u32(20)
    }
}

impl Syscall {
    pub fn parse(pid: Pid) -> Result<Syscall, SyscallParseError> {
        debug!("Parsing syscall entry...");
        wait_for_stop(pid)?;
        let regs = ptrace::getregs(pid)?;
        let args = SyscallArgs(regs.rdi, regs.rsi, regs.rdx, regs.r10, regs.r8, regs.r9);
        trace!("registers: {regs:?}");
        match regs.orig_rax {
            0 => {
                // TODO don't read it all
                let read = parse_return(pid, SyscallDisc::Read)?;
                let bytes = read_bytes_memory(pid, args.1 as usize, read as usize)?;
                Ok(Syscall::Read {
                    fd: args.0 as i32,
                    read_bytes: bytes,
                    requested_count: args.2 as usize,
                })
            }
            1 => {
                let text = read_bytes_memory(pid, args.1 as usize, args.2 as usize)?;
                trace!("address : {:#X}", args.1 as usize);
                parse_return(pid, SyscallDisc::Write)?;
                Ok(Syscall::Write {
                    fd: args.0 as i32,
                    buf: text,
                })
            }
            3 => {
                parse_return(pid, SyscallDisc::Close)?;
                Ok(Syscall::Close { fd: args.0 as i32 })
            }
            231 => Ok(Syscall::ExitGroup {
                status: args.0 as i32,
            }),
            257 => {
                trace!("doing openat");
                let pathname = read_cstring(pid, args.1 as usize)?;
                parse_return(pid, SyscallDisc::Openat)?;
                Ok(Syscall::Openat {
                    dirfd: args.0 as i32,
                    pathname,
                    flags: args.2 as i32,
                    mode: args.3 as u32,
                })
            }
            _ => {
                warn!("Unknown syscall was called");
                let return_value = get_return_value(pid)?;
                Ok(Syscall::Unknown { id: regs.orig_rax, args, return_value })
            }
        }
    }
}

pub struct SyscallIter(pub Pid);

impl Iterator for SyscallIter {
    type Item = Result<Syscall, SyscallParseError>;

    fn next(&mut self) -> Option<Self::Item> {
        match ptrace::syscall(self.0, None) {
            Err(Errno::ESRCH) => {
                error!("wtf");
                return None
            },
            Err(err) => return Some(Err(err.into())),
            _ => (),
        };
        match Syscall::parse(self.0) {
            Ok(call) => Some(Ok(call)),
            // TODO should process exit end the iterator?
            Err(err) => Some(Err(err)),
        }
    }
}
