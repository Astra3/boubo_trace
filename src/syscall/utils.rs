use std::io::IoSliceMut;

use libc::{c_void, RAX};
use log::{debug, info, trace};
use nix::{
    errno::Errno,
    sys::{
        ptrace,
        uio::{process_vm_readv, RemoteIoVec},
        wait::{waitpid, WaitStatus},
    },
    unistd::Pid,
};

use crate::syscall::SyscallParseError;

use super::SyscallDisc;

pub(super) fn wait_for_stop(pid: Pid) -> Result<(), SyscallParseError> {
    match waitpid(pid, None) {
        Ok(WaitStatus::Stopped(_, _)) => Ok(()),
        Ok(WaitStatus::Exited(_, exit_code)) => {
            info!("Process exited");
            Err(SyscallParseError::ProcessExit(exit_code))
        }
        Ok(status) => Err(SyscallParseError::UnexpectedWaitStatus(status)),
        Err(err) => Err(SyscallParseError::WaitPidError(err)),
    }
}

pub(super) fn read_bytes_memory(pid: Pid, base: usize, len: usize) -> Result<Vec<u8>, Errno> {
    let mut data = vec![0; len];
    process_vm_readv(
        pid,
        &mut [IoSliceMut::new(&mut data)],
        &[RemoteIoVec { base, len }],
    )?;
    Ok(data)
}

const MAX_BYTES_CSTRING: usize = 2048 * 1024; // 2 MiB
const BUFFER_SIZE: usize = 128;

// FIXME does this actually work?
pub(super) fn read_cstring(pid: Pid, base: usize) -> Result<Vec<u8>, Errno> {
    let mut data = vec![];
    let mut buf = [0; BUFFER_SIZE];
    let mut total_bytes_read = 0usize;

    while data.len() < MAX_BYTES_CSTRING {
        total_bytes_read += process_vm_readv(
            pid,
            &mut [IoSliceMut::new(&mut buf)],
            &[RemoteIoVec {
                base: base + total_bytes_read,
                len: BUFFER_SIZE,
            }],
        )?;

        if let Some(index) = buf.iter().position(|num| *num == 0) {
            data.extend_from_slice(&buf[..index + 1]);
            break;
        }
        data.extend_from_slice(&buf);
    }
    Ok(data)
}

pub(super) fn get_return_value(pid: Pid) -> Result<i64, SyscallParseError> {
    debug!("Parsing syscall exit...");
    ptrace::syscall(pid, None)?;
    wait_for_stop(pid)?;
    trace!("exit regs: {:?}", ptrace::getregs(pid)?);
    Ok(ptrace::read_user(pid, (RAX * 8) as *mut c_void)?)
}

pub(super) fn parse_syscall_error(return_value: i64) -> Errno {
    Errno::from_raw(-return_value as i32)
}

pub(super) fn parse_return(pid: Pid, syscall: SyscallDisc) -> Result<i64, SyscallParseError> {
    let ret = get_return_value(pid)?;
    trace!("return value: {ret}");
    if ret < 0 {
        return Err(SyscallParseError::SyscallError {
            syscall,
            error: parse_syscall_error(ret),
        });
    }
    Ok(ret)
}
