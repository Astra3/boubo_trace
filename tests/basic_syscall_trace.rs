#[cfg(test)]
mod tests {
    use std::{path::PathBuf, process::Command, sync::Once};

    use insta::glob;
    use nix::{
        sys::ptrace::{self, Options},
        unistd::Pid,
    };
    use spawn_ptrace::CommandPtraceSpawn;
    use tracer::syscall::{Syscall, SyscallIter, SyscallParseError};
    static INIT: Once = Once::new();

    fn initialize() {
        INIT.call_once(|| {
            println!("running init");
            let mut cmd = Command::new("make")
                .current_dir("test_programs/")
                .spawn()
                .unwrap();
            cmd.wait().unwrap();
        });
    }

    #[test]
    fn simple_test() {
        initialize();
        glob!("../test_programs/build/", "*.exec", |exec| {
            println!("path: {}", exec.display());
            let cmd = Command::new(exec)
                .current_dir(exec.parent().unwrap())
                .spawn_ptrace()
                .unwrap();
            let pid = Pid::from_raw(cmd.id() as i32);
            ptrace::setoptions(pid, Options::PTRACE_O_EXITKILL).unwrap();

            let it = SyscallIter(pid);
            // FIXME improve error serialize, if necessary
            let called_syscalls: Vec<_> = it
                .filter(|call| !matches!(call, Ok(Syscall::Unknown { .. })))
                .take_while(|call| !matches!(call, Err(SyscallParseError::ProcessExit(_))))
                .collect();
            insta::assert_debug_snapshot!(called_syscalls);
        });
    }
}
