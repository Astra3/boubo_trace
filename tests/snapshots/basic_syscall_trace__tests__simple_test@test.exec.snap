---
source: tests/basic_syscall_trace.rs
expression: called_syscalls
input_file: test_programs/build/test.exec
snapshot_kind: text
---
[
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Err(
        SyscallError {
            syscall: Openat,
            error: ENOENT,
        },
    ),
    Ok(
        Openat {
            dirfd: -100,
            pathname: [
                47,
                101,
                116,
                99,
                47,
                108,
                100,
                46,
                115,
                111,
                46,
                99,
                97,
                99,
                104,
                101,
                0,
            ],
            flags: 524288,
            mode: 0,
        },
    ),
    Ok(
        Close {
            fd: 3,
        },
    ),
    Ok(
        Openat {
            dirfd: -100,
            pathname: [
                47,
                117,
                115,
                114,
                47,
                108,
                105,
                98,
                47,
                108,
                105,
                98,
                99,
                46,
                115,
                111,
                46,
                54,
                0,
            ],
            flags: 524288,
            mode: 0,
        },
    ),
    Ok(
        Read {
            fd: 3,
            read_bytes: [
                127,
                69,
                76,
                70,
                2,
                1,
                1,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                3,
                0,
                62,
                0,
                1,
                0,
                0,
                0,
                224,
                95,
                2,
                0,
                0,
                0,
                0,
                0,
                64,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                120,
                173,
                30,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                0,
                56,
                0,
                14,
                0,
                64,
                0,
                63,
                0,
                62,
                0,
                6,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                64,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                8,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                3,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                128,
                175,
                27,
                0,
                0,
                0,
                0,
                0,
                128,
                175,
                27,
                0,
                0,
                0,
                0,
                0,
                128,
                175,
                27,
                0,
                0,
                0,
                0,
                0,
                30,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                30,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                1,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                176,
                50,
                2,
                0,
                0,
                0,
                0,
                0,
                176,
                50,
                2,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                1,
                0,
                0,
                0,
                5,
                0,
                0,
                0,
                0,
                64,
                2,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                2,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                2,
                0,
                0,
                0,
                0,
                0,
                57,
                4,
                23,
                0,
                0,
                0,
                0,
                0,
                57,
                4,
                23,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                1,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                0,
                80,
                25,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                25,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                25,
                0,
                0,
                0,
                0,
                0,
                44,
                221,
                4,
                0,
                0,
                0,
                0,
                0,
                44,
                221,
                4,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                1,
                0,
                0,
                0,
                6,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                80,
                75,
                0,
                0,
                0,
                0,
                0,
                0,
                64,
                208,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                2,
                0,
                0,
                0,
                6,
                0,
                0,
                0,
                64,
                105,
                30,
                0,
                0,
                0,
                0,
                0,
                64,
                105,
                30,
                0,
                0,
                0,
                0,
                0,
                64,
                105,
                30,
                0,
                0,
                0,
                0,
                0,
                48,
                2,
                0,
                0,
                0,
                0,
                0,
                0,
                48,
                2,
                0,
                0,
                0,
                0,
                0,
                0,
                8,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                80,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                8,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                160,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                160,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                160,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                68,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                68,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                7,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                136,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                8,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                83,
                229,
                116,
                100,
                4,
                0,
                0,
                0,
                80,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                3,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                8,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                80,
                229,
                116,
                100,
                4,
                0,
                0,
                0,
                160,
                175,
                27,
                0,
                0,
                0,
                0,
                0,
                160,
                175,
                27,
                0,
                0,
                0,
                0,
                0,
                160,
                175,
                27,
                0,
                0,
                0,
                0,
                0,
                124,
                120,
                0,
                0,
                0,
                0,
                0,
                0,
                124,
                120,
                0,
                0,
                0,
                0,
                0,
                0,
                4,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                81,
                229,
                116,
                100,
                6,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                16,
                0,
                0,
                0,
                0,
                0,
                0,
                0,
                82,
                229,
                116,
                100,
                4,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                120,
                59,
                30,
                0,
                0,
                0,
                0,
                0,
                136,
                52,
                0,
                0,
                0,
                0,
                0,
                0,
            ],
            requested_count: 832,
        },
    ),
    Ok(
        Close {
            fd: 3,
        },
    ),
    Ok(
        Write {
            fd: 1,
            buf: [
                72,
                101,
                108,
                108,
                111,
                32,
                119,
                111,
                114,
                108,
                100,
                33,
                10,
            ],
        },
    ),
    Ok(
        ExitGroup {
            status: 0,
        },
    ),
]
